core/transformer/src/IModelTransformer.ts

```diff
@@ -1084,6 +1092,154 @@ export class IModelTransformer extends IModelExportHandler {
+    });
+    const lastProvenanceEntityInfo: IModelTransformer["_lastProvenanceEntityInfo"] = db.withSqliteStatement(
+      `SELECT entityId, aspectId, aspectVersion, aspectKind FROM ${IModelTransformer.lastProvenanceEntityInfoTable}`,
+      (stmt) => {
+        if (DbResult.BE_SQLITE_ROW !== stmt.step())
+          throw Error(
+            "expected row when getting lastProvenanceEntityId from target state table"
+          );
+        return {
+          entityId: stmt.getValueId(0),
+          aspectId: stmt.getValueId(1),
+          aspectVersion: stmt.getValueString(2),
+          aspectKind: stmt.getValueString(3) as ExternalSourceAspect.Kind,
+        };
+      }
+    );
+    const targetHasCorrectLastProvenance = transformer.provenanceDb.withPreparedStatement(`
+      SELECT Version FROM ${ExternalSourceAspect.classFullName}
+      WHERE Scope.Id=:scopeId
+        AND ECInstanceId=:aspectId
+        AND Kind=:kind
+        AND Element.Id=:entityId
+    `,
+    (statement: ECSqlStatement): boolean => {
+      statement.bindId("scopeId", transformer.targetScopeElementId);
+      statement.bindId("aspectId", lastProvenanceEntityInfo.aspectId);
+      statement.bindString("kind", lastProvenanceEntityInfo.aspectKind);
+      statement.bindId("entityId", lastProvenanceEntityInfo.entityId);
+      const stepResult = statement.step();
+      switch (stepResult) {
+        case DbResult.BE_SQLITE_ROW:
+          const version = statement.getValue(0).getString();
+          return version === lastProvenanceEntityInfo.aspectVersion;
+        case DbResult.BE_SQLITE_DONE:
+          return false;
+        default:
+          throw new IModelError(IModelStatus.SQLiteError, `got sql error ${stepResult}`);
+      }
+    });
+    if (!targetHasCorrectLastProvenance)
+      throw Error([
+        "Target for resuming from does not have the expected provenance ",
+        "from the target that the resume state was made with",
+      ].join("\n"));
+    transformer._lastProvenanceEntityInfo = lastProvenanceEntityInfo;
+    // force assign to readonly options since we do not know how the transformer subclass takes options to pass to the superclass
+    (transformer as any)._options = state.options;
+    transformer.importer.loadStateFromJson(state.importerState);
+    transformer.exporter.loadStateFromJson(state.exporterState);
+    transformer.context.loadStateFromDb(db);
+    db.closeDb();
+    return transformer as InstanceType<SubClass>;
+  }
+
+  /**
+   * Serialize the state of the active transformation to a file path, if a file at the path already exists, it will be overwritten
+   * This state can be used by [[resumeTransformation]] to resume a transformation from this point.
+   * The serialization format is a custom sqlite database
+   * @note custom transformers with custom state may need to override this method in order to handle storing their own custom state somewhere,
```

Comment:
How likely is it that custom transformers will indeed need to override this? It looks awkward to do so since the base implementation is very self-contained (create db, write to it, close it, return nothing).
Also is there any risk of one custom transformer finding a state db leftover from a different custom transformer, attempting to resume, and being only partially constructed?

Reply:
It is certainly significantly likely. They don't have to store it in the database itself, but I think because I'm anticipating people wanting to do that to have one self contained resumable dump, I am going to add a new protected method with access to the `SqliteDb` object, which is going to be the override point. Please let me know if you have any concerns with that approach.

As for attempting to resume a dump from the wrong class, I will add an additional item to the json state to id the transformer  class used to make the dump. That can be used to sanity check. Thank you for pointing it out.

Reply:
I have added tests demonstrating failing on attempting to resume from a different importer, exporter, or transformer class, based on the name of the class/constructor function which is now saved to the dump
.
I also added a few new `additionalData` functions for making it easier for custom transformers to store custom state without accessing the sqlitedb, and made it easier for them to access the sqlitedb in case they need to store it there. This has a side effect of cleaning up the code a tiny bit. There is also a test demonstrating that resuming does in fact pick up additional data across all the added new functions.